contracts/StorageAccessible.sol

```diff
@@ -52,20 +52,47 @@ contract StorageAccessible {
      */
     function simulateDelegatecall(
         address targetContract,
-        bytes memory calldataPayload
+        bytes calldata calldataPayload
     ) public returns (bytes memory response) {
-        bytes memory innerCall = abi.encodeWithSelector(
-            this.simulateDelegatecallInternal.selector,
-            targetContract,
-            calldataPayload
-        );
-        (, response) = address(this).call(innerCall);
-        bool innerSuccess = response[response.length - 1] == 0x01;
-        setLength(response, response.length - 1);
-        if (innerSuccess) {
-            return response;
-        } else {
-            revertWith(response);
+        bytes4 selector = this.simulateDelegatecallInternal.selector;
+        assembly {
+            let internalCalldata := mload(0x40)
+            mstore(internalCalldata, selector)
+            // Abuse the fact that both this and the internal methods have the
+            // same signature, and differ only in symbol name (and therefore,
+            // selector) and copy calldata directly. This saves us approximately
+            // 250 bytes of code and 300 gas at runtime over the
+            // `abi.encodeWithSelector` builtin.
+            calldatacopy(
+                add(internalCalldata, 0x04),
+                0x04,
+                sub(calldatasize(), 0x04)
+            )
+
+            pop(call(
+                gas(),
+                address(),
+                0,
+                internalCalldata,
+                calldatasize(),
+                // Store the `success` value at memory address 0x00 (reserved
+                // Solidity scratch space).
+                0x00,
```

Comment:
I don't think this logic is correct. It might work with our tests as we never revert with an error message. As far as I understand https://docs.soliditylang.org/en/v0.8.3/yul.html#yul-call-return-area the return data is really only the data (so the data returned with the opcodes `return` or `revert`), but no the success flag.

Reply:
nvm ... I should read until the end xD

Reply:
I'll expand the comment so its more obvious.